Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift

```diff
@@ -0,0 +1,224 @@
+  let noteDescriptors: [NoteDescriptor]
+
+  /// Descriptors for any Fix-Its that can be applied for this diagnostic message.
+  let fixIts: [FixIt]  // TODO: How to create an abstract model for this?
+
+  /// Initializes a new `DiagnosticDescriptor`.
+  ///
+  /// - Parameters:
+  ///   - locationMarker: The marker pointing to location in source code.
+  ///   - id: The message ID of the diagnostic.
+  ///   - message: The textual message to display for the diagnostic.
+  ///   - severity: The severity level of the diagnostic. Default is `.error`.
+  ///   - highlight: The syntax elements to be highlighted. Default is an empty array.
+  ///   - noteDescriptors: An array of note descriptors for additional context. Default is an empty array.
+  ///   - fixIts: An array of Fix-It descriptors for quick fixes. Default is an empty array.
+  init(
+    locationMarker: LocationMarker,
+    id: MessageID = MessageID(domain: "test", id: "conjured"),
+    message: String,
+    severity: DiagnosticSeverity = .error,
+    highlight: [Syntax] = [],
+    noteDescriptors: [NoteDescriptor] = [],
+    fixIts: [FixIt] = []
+  ) {
+    self.locationMarker = locationMarker
+    self.id = id
+    self.message = message
+    self.severity = severity
+    self.highlight = highlight
+    self.noteDescriptors = noteDescriptors
+    self.fixIts = fixIts
+  }
+
+  /// Creates a ``Diagnostic`` instance from a given ``DiagnosticDescriptor``, syntax tree, and location markers.
+  ///
+  /// - Parameters:
+  ///   - tree: The syntax tree where the diagnostic is rooted.
+  ///   - markers: A dictionary mapping location markers to their respective offsets in the source code.
+  ///
+  /// - Throws:
+  ///   - Error if the location marker is not found in the source code.
+  ///   - Error if a node corresponding to a given marker is not found in the syntax tree.
+  ///
+  /// - Returns: A ``Diagnostic`` instance populated with details from the ``DiagnosticDescriptor``.
+  func createDiagnostic(
+    inSyntaxTree tree: some SyntaxProtocol,
+    usingLocationMarkers markers: [LocationMarker: Int]
+  ) throws -> Diagnostic {
+    func node(at marker: LocationMarker) throws -> Syntax {
+      guard let markedOffset = markers[marker] else {
+        throw DiagnosticCreationError(message: "Marker \(marker) not found in the marked source")
+      }
+      let markedPosition = AbsolutePosition(utf8Offset: markedOffset)
+      guard let token = tree.token(at: markedPosition) else {
+        throw DiagnosticCreationError(message: "Node not found at marker \(marker)")
+      }
+      return Syntax(token)
+    }
+
+    let diagnosticNode = try node(at: self.locationMarker)
+
+    let notes = try self.noteDescriptors.map { noteDescriptor in
+      let noteNode = try node(at: noteDescriptor.locationMarker)
+
+      return Note(
+        node: noteNode,
+        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
+      )
```

Comment:
Nit: `noteNode` is a weird name and we don’t actually need it.

```suggestion
      return Note(
        node: try node(at: noteDescriptor.locationMarker),
        message: SimpleNoteMessage(message: noteDescriptor.message, fixItID: noteDescriptor.id)
      )
```

Reply:
Great idea! Done